File: frontend/lib/accountService.ts
import executionApi from "./executionApi";

export interface AccountResponse {
  id: string;
  userId: string;
  role: string;
  accountNumber: string;
  balance: number;
  pendingBalance: number;
  createdAt: string;
}

export interface BankAccountResponse {
  id: string;
  bankName: string;
  accountNumber: string;
  accountHolderName: string;
  defaultAccount: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface BankAccountRequest {
  bankName: string;
  accountNumber: string;
  accountHolderName: string;
}

export const accountService = {
  async getAccount(): Promise<AccountResponse> {
    const res = await executionApi.get<AccountResponse>("/api/account");
    return res.data;
  },

  async deposit(projectId: string, amount: number, bankAccountId: string): Promise<AccountResponse> {
    const res = await executionApi.post<AccountResponse>("/api/account/deposit", {
      projectId,
      bankAccountId,
      amount,
    });
    return res.data;
  },

  async getBankAccounts(): Promise<BankAccountResponse[]> {
    const res = await executionApi.get<BankAccountResponse[]>("/api/account/bank-accounts");
    return res.data;
  },

  async addBankAccount(data: BankAccountRequest): Promise<BankAccountResponse> {
    const res = await executionApi.post<BankAccountResponse>("/api/account/bank-accounts", data);
    return res.data;
  },

  async updateBankAccount(id: string, data: BankAccountRequest): Promise<BankAccountResponse> {
    const res = await executionApi.put<BankAccountResponse>(`/api/account/bank-accounts/${id}`, data);
    return res.data;
  },

  async deleteBankAccount(id: string): Promise<void> {
    await executionApi.delete(`/api/account/bank-accounts/${id}`);
  },

  async setDefaultBankAccount(id: string): Promise<void> {
    await executionApi.patch(`/api/account/bank-accounts/${id}/default`);
  },
};

File: backend/ProjectWorker-Execution-Service/src/main/java/com/example/ProjectWorker_Execution_Service/service/AccountService.java
package com.example.ProjectWorker_Execution_Service.service;

import com.example.ProjectWorker_Execution_Service.dto.AccountResponse;
import com.example.ProjectWorker_Execution_Service.dto.BankAccountRequest;
import com.example.ProjectWorker_Execution_Service.dto.BankAccountResponse;
import com.example.ProjectWorker_Execution_Service.dto.DepositRequest;
import com.example.ProjectWorker_Execution_Service.security.UserPrincipal;

import java.util.List;

public interface AccountService {

    AccountResponse getOrCreateAccount(UserPrincipal principal);

    AccountResponse deposit(DepositRequest request, UserPrincipal principal);

    List<BankAccountResponse> getBankAccounts(UserPrincipal principal);

    BankAccountResponse addBankAccount(BankAccountRequest request, UserPrincipal principal);

    BankAccountResponse updateBankAccount(String id, BankAccountRequest request, UserPrincipal principal);

    void deleteBankAccount(String id, UserPrincipal principal);

    void setDefaultBankAccount(String id, UserPrincipal principal);
}
